### ISSUE TITLE
Move number formatting code to Plurimath

### ISSUE THREAD

**Issue by opoudjis** [metanorma/isodoc#571]
In https://github.com/metanorma/iso-iec-80000/issues/39 the desire was expressed to move number formatting code from Metanorma to Plurimath. The code is 

https://github.com/metanorma/isodoc/blob/main/lib/isodoc/presentation_function/math.rb : localize_maths(), and probably twitter_cldr_localiser() (which draws on [twitter-cldr-rb](https://github.com/twitter/twitter-cldr-rb) for its definition of number localisation, and expands on it).

Note that flavours customise twitter_cldr_localiser_symbols(), to define default rendering for the Metanorma flavour, overriding the values given by twitter-cldr-rb for the current locale. For example,

https://github.com/metanorma/isodoc/blob/2d650f7788abf0eb14d89858c7877c3e36f927d3/lib/isodoc/presentation_function/math.rb#L65

````
def twitter_cldr_localiser_symbols
        { group: "'" }
end
````

That tells ITU to break up groups of three digits with "'" not space.

https://github.com/metanorma/metanorma-bipm/blob/4aab355394d6a6b18ff884100627f3d102548ffb/lib/isodoc/bipm/presentation_xml_convert.rb#L77

````
      def twitter_cldr_localiser_symbols
        { group: "&#xA0;", fraction_group: "&#xA0;",
          fraction_group_digits: 3 }
      end
````

That tells BIPM to separate digits and fraction digits, every 3 digits, with nbsp. I would want to pass the hash in twitter_cldr_localiser_symbols to the class realising this formatter, since they are defaults for the entire run.

I am about to do some refactoring of associated code, triggered by https://github.com/metanorma/metanorma-standoc/issues/882 , but once that settles down, I will request that you take this code on to Plurimath, so long as I can call it cleanly: it will be a different class of code than the Plurimath code conversions, and it will take a numeric string as input, not a MathML node. Note that localize_maths() in that code is actually iterating through mn, and calculating what the precision of the number is if it isn't given. I would instead give you just the string, and ask you to calculate the precision if not given:

````
def localize_maths(node, locale)
      node.xpath(".//m:mn", MATHML).each do |x|
        num = BigDecimal(x.text)
        precision = /\./.match?(x.text) ? x.text.sub(/^.*\./, "").size : 0
        x.children = localized_number(num, locale, precision)
      rescue ArgumentError
      end
    end
````

I would want you to realise:

````
class Plurimath::Numberformatter
def initialize(locale_str: "en", twitter_cldr_localiser_symbols: {})
  @locale,  @twitter_cldr_reader = twitter_cldr_localiser(locale_str)
  @twitter_cldr_localiser_symbols = twitter_cldr_localiser_symbols
end

    def twitter_cldr_reader(locale)
      return @twitter_cldr_reader if @twitter_cldr_reader

      num = TwitterCldr::DataReaders::NumberDataReader.new(locale)
      num.symbols.merge!(@twitter_cldr_localiser_symbols)
        .merge!(parse_localize_number)
    end

    def twitter_cldr_localiser(locale_str: "en")
      locale = TwitterCldr.supported_locale?(locale_str.to_sym) ? locale_str.to_sym : :en
      [locale , twitter_cldr_reader(locale)]
    end

  def parse_localize_number
...
 end

def localized_number(number_string, locale_str: "en", precision: nil)
   locale, reader = if locale_str == @locale.to_s
     [@locale,  @twitter_cldr_reader]
   else
     twitter_cldr_reader(locale_str)
   end
   num = BigDecimal(number_string.to_s)
   precision ||= /\./.match?(x.text) ? x.text.sub(/^.*\./, "").size : 0
   localized_number(num, locale, reader, precision) # the existing implementation, but with twitter_cldr_reader now an explicit parameter, in case we switch it mid-run
end

end
````

So that I can call: 

````
n = Plurimath::Numberformatter.new()
n.localized_number(1)
localized_number("1")
localized_number(1.6, "en")
localized_number(1.6, "fr", 3)
````

BUT. Do not do anything yet, this code is going to be unstable for a little while.

**Comment by ronaldtse** [comment]
Let’s not forget that there are two cases:
1. Pure numbers to be localized as text
2. Numbers to be localized within MathML formulas. 

They are likely share some functionality but the output (localized number vs localized number in MathML) is very different.

**Comment by ronaldtse** [comment]
And actually I’m not sure if we should read anything from CLDR since they ignore both ISO 80000 and BIPM number formats. 

In fact, we need our own Numbers model in Plurimath if we need to do MathML Content, so might as well implement our own numbers string formatter.

**Comment by opoudjis** [comment]
> Let’s not forget that there are two cases:
> 
> 1. Pure numbers to be localized as text
> 2. Numbers to be localized within MathML formulas.
> 
> They are likely share some functionality but the output (localized number vs localized number in MathML) is very different.

That's why I want just the former implemented. The latter will involve recursively calling the former for all `mn`. And I explicitly made that distinction above, between `localize_maths` and `localize_number`.

**Comment by opoudjis** [comment]
> And actually I’m not sure if we should read anything from CLDR since they ignore both ISO 80000 and BIPM number formats.
> 
> In fact, we need our own Numbers model in Plurimath if we need to do MathML Content, so might as well implement our own numbers string formatter.

BIPM and ISO are not the only standards we support, nor are French and English the only languages we support. We need a fallback number formatter when the SDO doesn't specify one, and that is what CLDR gives us. CLDR also gives us the parameters for number formatting, which we are expanding.

**Comment by opoudjis** [comment]
I'm now thinking that, with so many tickets coming up, I should leave the implementation of scientific notation formatting to @suleman-uzair . But @ronaldtse Please provide me with the expected formatting for ISO 2355, so I can include that in the spec of behaviour I pass over.

**Comment by opoudjis** [comment]
1.27783 +3 = 1277.83
1.47816 +0 = 1.47816

... I guess I can spec this in a template, but it's eccentric....

.... #.#e+0, with + meaning indicate the sign as positive or negative, and a separate config to tell it that the rendering of the exponent letter is space, not "e".

**Comment by opoudjis** [comment]
I think this can move to Plurimath, I'll just tidy this up before creating a new ticket there.

**Comment by opoudjis** [comment]
Action: rewrite this ticket, move to Plurimath. Once Plurimath implements, use that method in isodoc instead of current code.

**Comment by ronaldtse** [comment]
@opoudjis I don't think we should implement the funny exponential format because it can only be chosen based on a "group of numbers" in the column. It's a manual choice. This is too arbitrary.

What is reasonable though is the display of "number of digits", say "n", i.e. fit this number with as many significant digits placed into n-digits.

e.g.
```
fit_digits(0.000000111222333, total: 6) => "1.11222 -7"
fit_digits(0.000000111222333, decimal: 6) => "1.112223 -7"
```

**Comment by opoudjis** [comment]
> @opoudjis I don't think we should implement the funny exponential format because it can only be chosen based on a "group of numbers" in the column. It's a manual choice. This is too arbitrary.

You are referring to "1.27783 +3"? ... What is arbitrary about it? That the number of significant digits goes from 6 to 5 if the exponent is greater than 9? 

> What is reasonable though is the display of "number of digits", say "n", i.e. fit this number with as many significant digits placed into n-digits.
> 
> e.g.
> 
> ```
> fit_digits(0.000000111222333, total: 6) => "1.11222 -7"
> fit_digits(0.000000111222333, decimal: 6) => "1.112223 -7"
> ```

So, differentiating precision, the decimal digits, from the overall digits in the mantissa? I mean, we can, but won't they always differ by 1?

**Comment by ronaldtse** [comment]
> You are referring to "1.27783 +3"? ... What is arbitrary about it? That the number of significant digits goes from 6 to 5 if the exponent is greater than 9?

Look at this:
<img width="402" alt="Screenshot 2024-04-02 at 14 39 41" src="https://github.com/metanorma/isodoc/assets/11865/0a499f8a-0a9a-45d0-a741-335324464b0d">

I'm not sure if we can automate this notation across table cells easily.

**Comment by opoudjis** [comment]
Pending bug fix in Plurimath

**Comment by opoudjis** [comment]
metanorma-bipm uses the Twitter CLDR characters of the current locale to do post-processing. This means these now need to be exposed by NumberFormatter in Plurimath, since metanorma-bipm need to know the current locale's decimal point symbol, and it should not have to re-open twitter-cldr independently.

**Comment by opoudjis** [comment]
Defining `twitter_cldr_reader` method on Plurimath NumberFormatter

**Comment by opoudjis** [comment]
And needed to update bipm to use the now exposed Twitter CLDR characters
